### ISSUE TITLE
texture submit; need a way to specify target

### ISSUE THREAD

**Issue by kevinresol** [underscorediscovery/luxe#421]
**TL; DR;**

In summary, what changes Luxe needs is:

1. Add GL_TEXTURE_EXTERNAL_OES to [TextureType](https://github.com/underscorediscovery/luxe/blob/f1db62d/phoenix/Texture.hx#L447) and [TextureSubmitTarget](https://github.com/underscorediscovery/luxe/blob/f1db62d/phoenix/Texture.hx#L439)
2. Allow specifying submit target from outside (through options), or intelligently decide by itself instead of [just throwing](https://github.com/underscorediscovery/luxe/blob/f1db62d/phoenix/Texture.hx#L173)

**Long version:**

I am working on android, using `SurfaceTexture` from the android API. Its `updateTexImage` method renders to the `GL_TEXTURE_EXTERNAL_OES` texture target (instead of `GL_TEXTURE_2D`).

I first created a `RenderTexture` in Luxe, then passes the texture name (id) to Java and uses the said API to render some images into the texture. In order for the thing to work, I need to:

1. Set the RenderTexture type to `GL_TEXTURE_EXTERNAL_OES` (`0x8D65`)
2. Make the RenderTexture submit to `GL_TEXTURE_EXTERNAL_OES`
3. The fragment shader of the Visual that renders the RenderTexture should use `samplerExternalOES` instead of `sampler2D`

But since there seems no way to control the submit target from outside, (every call to [`submit`](https://github.com/underscorediscovery/luxe/blob/f1db62d/phoenix/Texture.hx#L159) has omitted the `_target` argument), I resorted to hardcoding the required target into the function.

**Comment by underscorediscovery** [comment]
Hmm, first off - the TextureType and TextureSubmitTarget function as `Int`, meaning you can use `GL_TEXTURE_EXTERNAL_OES (0x8D65)` in place of a typed value as an Int and it will work just fine. Secondly, submit is only called if you are giving the texture some pixels on create. You haven't described this use case.

Call submit yourself with the right arguments, they don't have to be part of the enum to be passed in.

But the question is, are you calling `new` every time or something?

**Comment by kevinresol** [comment]
ok, I didn't investigate thoroughly before posting. Here it is "real problem"

I created an RenderTexture like this:

```haxe
renderTexture = new RenderTexture({
	id: 'video_player_texture',
	width: 512,
	height: 512,
        type: 0x8D65, // GL_TEXTURE_EXTERNAL_OES
});
```

In the constructor of RenderTexture, it [calls submit(null)](https://github.com/underscorediscovery/luxe/blob/e374019/phoenix/RenderTexture.hx#L41)

And then fails [here](https://github.com/underscorediscovery/luxe/blob/e374019/phoenix/Texture.hx#L173)

Basically the [`options.texture == null` check](https://github.com/underscorediscovery/luxe/blob/e374019/phoenix/RenderTexture.hx#L40) is useless, because one is always created in the `super()` call.

And there is no way to make it call submit with a the correct target, which is the problem.

**Comment by underscorediscovery** [comment]
right so, the texture handle needs to exist, so the call to super is ok. Call submit yourself immediately after new?

**Comment by underscorediscovery** [comment]
I'm trying to understand the actual problem you see, not the conceptual issue in the code, but the side effect of it.

**Comment by kevinresol** [comment]
The actual problem is, write this piece of code, then it crashes.

```haxe
new RenderTexture({
	id: 'rt',
	width: 512,
	height: 512,
	type: 0x8D65, // GL_TEXTURE_EXTERNAL_OES
});
```

**Comment by underscorediscovery** [comment]
I tidied up the submit in [this branch](https://github.com/underscorediscovery/luxe/tree/texture-submit) but since I can't test it with that texture type you'll have to test

**Comment by kevinresol** [comment]
Sure I will test it. Thanks!
But I also wonder isn't adding a `submitTarget` field to TextureOption more convenient?

**Comment by underscorediscovery** [comment]
Not really, the type is meant to be inferred from the TextureType, and TextureType is supposed to be (but isn't yet) an abstraction, which would mean that having GL concepts hardcoded in the API which I was trying to avoid

**Comment by kevinresol** [comment]
> I tidied up the submit in this branch but since I can't test it with that texture type you'll have to test

Yes I can confirm it works now
